src/dataflow/context.rs

```diff
@@ -1,11 +1,72 @@
-use std::marker::PhantomData;
+use std::{collections::HashMap, iter::Iterator, marker::PhantomData, sync::Arc};
 
 use serde::Deserialize;
 
 use crate::dataflow::{
-    operator::OperatorConfig, AppendableStateT, Data, StateT, Timestamp, WriteStream,
+    deadlines::{DeadlineId, DeadlineT},
+    operator::OperatorConfig,
+    stream::StreamId,
+    AppendableStateT, Data, StateT, Timestamp, WriteStream,
 };
 
+/*************************************************************************************************
+ * SetupContext: Provided to an operator's `setup` method, and allows the operators to register  *
+ * deadlines.                                                                                    *
+ ************************************************************************************************/
+
+/// A `SetupContext` is made available to an operator's `setup` method, and allows the operators to
+/// register deadlines for events along with their corresponding handlers.
+pub struct SetupContext<S> {
+    deadlines: HashMap<DeadlineId, Arc<dyn DeadlineT<S>>>,
+    // TODO (Sukrit): Can we provide a better interface than ReadStream and WriteStream IDs?
+    read_stream_ids: Vec<StreamId>,
+    write_stream_ids: Vec<StreamId>,
+}
+
+impl<S> SetupContext<S> {
+    pub fn new(read_stream_ids: Vec<StreamId>, write_stream_ids: Vec<StreamId>) -> Self {
+        Self {
+            deadlines: HashMap::new(),
+            read_stream_ids,
+            write_stream_ids,
+        }
+    }
+
+    /// Register a deadline with the system.
+    pub fn add_deadline(&mut self, deadline: impl DeadlineT<S> + 'static) {
+        let deadline_id = deadline.get_id();
+        self.deadlines.insert(deadline_id, Arc::new(deadline));
+    }
+
+    /// Get the deadlines registered in this context.
+    pub(crate) fn get_deadlines(&mut self) -> impl Iterator<Item = &mut Arc<dyn DeadlineT<S>>> {
+        self.deadlines.values_mut()
+    }
+
+    /// Get the identifiers of the read streams of this operator.
+    pub fn get_read_stream_ids(&self) -> &Vec<StreamId> {
+        &self.read_stream_ids
+    }
+
+    /// Get the identifiers of the write streams of this operator.
+    pub fn get_write_stream_ids(&self) -> &Vec<StreamId> {
+        &self.write_stream_ids
+    }
```

Comment:
Are these used internally or do these need to be user accessible?

If they're only used internally, can we make them `pub(crate)`?

Reply:
Yup, they're internal only. Added `pub(crate)`